### PR TITLE
[babel-plugin] add "last media query" wins behaviour behind feature flag

### PR DESCRIPTION
**Pull request by mellyeliu**
combing thru the h1 stragglers...

Context: 
- in https://github.com/facebook/stylex/pull/898 we added a media query transformer to implement "last media query" wins behaviour which is expected amongst devs but didn't integrate it into the babel plugin
- previously devs would have to manually ensure this behaviour, but it's increasingly unsustainable esp with `defineConsts`/ shareable media queries

`media-query-transformer` fixes:
- reduced dfs depth by one, as by the time we're in `flatten-raw-style-obj.js` the namespace has been stripped
- also added an array check to avoid messing with `firstThatWorks` arrays

added in `babel-plugin`
- add `style-value-parser` as dependency
- added processing in `flatten-raw-style-obj.js` and update tests
- gated behind `enableLastMediaQueryWins` param

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -36,6 +38,7 @@ jobs:
         with:
           node-version: '22.x'
       - run: npm install
+      - run: npm run build --workspaces --if-present
```
Comment (necolas) [inline .github/workflows/tests.yml]: This already runs after install https://github.com/facebook/stylex/blob/main/package.json#L7

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
   "scripts": {
     "postinstall": "npm run build",
     "prepare": "husky install tools/husky",
-    "build": "npm run build --workspaces --if-present",
+    "build": "npm run build -w style-value-parser && npm run build --workspaces --if-present",
```
Comment (necolas) [inline package.json]: If we need the parser package built first it can be listed first in the workspaces array, like we did with `shared`

File: packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js [hunk 3]
```diff
@@ -1829,12 +1830,71 @@ describe('@stylexjs/babel-plugin', () => {
...
+              root: {
+                backgroundColor: {
+                  default: 'red',
+                  '@media (min-width: 1000px)': 'blue',
+                  '@media (min-width: 2000px)': 'purple',
+                }
+              },
+            });
+          `,
+            { enableLastMediaQueryWins: true },
```
Comment (necolas) [inline packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js]: Is this test meant to have the flag enabled?

File: packages/@stylexjs/babel-plugin/package.json [hunk 4]
```diff
@@ -22,7 +22,8 @@
     "@babel/types": "^7.26.8",
     "@dual-bundle/import-meta-resolve": "^4.1.0",
     "@stylexjs/stylex": "0.14.2",
-    "postcss-value-parser": "^4.1.0"
+    "postcss-value-parser": "^4.1.0",
+    "style-value-parser": "*"
```
Comment (necolas) [inline packages/@stylexjs/babel-plugin/package.json]: This should probably be pinned to the current version like we do with other packages

File: packages/style-value-parser/package.json [hunk 5]
```diff
@@ -1,23 +1,29 @@
 {
-  "private": true,
   "name": "style-value-parser",
   "version": "0.14.2",
-  "main": "src/index.js",
+  "private": true,
```
Comment (necolas) [inline packages/style-value-parser/package.json]: Prefer private packages to have this field first so it's immediately obvious they are private

File: packages/style-value-parser/package.json [hunk 6]
```diff
@@ -1,23 +1,29 @@
...
+  "type": "module",
+  "main": "lib/index.js",
   "license": "MIT",
   "scripts": {
+    "prepublishOnly": "npm run build",
```
Comment (necolas) [inline packages/style-value-parser/package.json]: Why is this needed?

**Comment by mellyeliu** [comment]
sorry this is a draft, will mark when ready

File: packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js [hunk 7]
```diff
@@ -1817,8 +1817,66 @@ describe('@stylexjs/babel-plugin', () => {
...
+                  {
+                    "ltr": ".xrkmrrc{background-color:red}",
+                    "rtl": null,
+                  },
+                  3000,
+                ],
+                [
+                  "x1wa28t1",
+                  {
+                    "ltr": "@media (max-width: 1000px) and (not (max-width: 500px)){.x1wa28t1.x1wa28t1{background-color:blue}}",
```
Comment (nmn) [inline packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js]: This should be simplified here.

**Comment by mellyeliu** [comment]
let me write out a merge intervals approach for min/max simplification

File: package.json [hunk 8]
```diff
@@ -39,7 +40,7 @@
     "jest": "^30.0.0-alpha.7",
     "jest-environment-jsdom": "^30.0.0-alpha.7",
     "lint-staged": "^13.0.3",
-    "prettier": "^3.5.0",
+    "prettier": "^3.2.5",
```
Comment (mellyeliu) [inline package.json:43]: this version of prettier was causing some parsing issues, after quick search I pinned it to this which seemed more stable

**Comment by mellyeliu** [comment]
graveyard of force pushes from debugging workflow/rebase struggles...

File: packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js [hunk 9]
```diff
@@ -1817,8 +1817,75 @@ describe('@stylexjs/babel-plugin', () => {
...
+                backgroundColor: {
+                  default: 'red',
+                  '@media (max-width: 900px)': 'blue',
+                  '@media (max-width: 500px)': 'purple',
+                  '@media (max-width: 400px)': 'green',
+                }
+              },
+            });
+          `,
+            { enableLastMediaQueryWins: true },
```
Comment (nmn) [inline packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js:1835]: We might need a better name for this.

File: packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js [hunk 10]
```diff
@@ -43,30 +39,26 @@ describe('Media Query Transformer', () => {
...
-            '@media (max-width: 768px)': '1 / -1',
-          },
+      gridColumn: {
+        default: '1 / 2',
+        '@media (min-width: 1024.01px) and (max-width: 1440px)': {
           '@media (min-width: 768.01px) and (max-width: 1024px)': '1 / 3',
           '@media (max-width: 768px)': '1 / -1',
         },
+        '@media (min-width: 768.01px) and (max-width: 1024px)': '1 / 3',
+        '@media (max-width: 768px)': '1 / -1',
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js]: The media queries can be flattened in this case, but this is a contrived example. Using `height` queries in the nested object is probably more useful.

File: packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js [hunk 11]
```diff
@@ -16,24 +16,20 @@ const stylex = {
...
     const originalStyles = {
-      foo: {
-        gridColumn: {
-          default: '1 / 2',
-          '@media (max-width: 1440px)': '1 / 4',
-          '@media (max-width: 1024px)': '1 / 3',
-          '@media (max-width: 768px)': '1 / -1',
-        },
+      gridColumn: {
+        default: '1 / 2',
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js:16]: Please add examples which use a series of `min-width` instead.

File: packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js [hunk 12]
```diff
@@ -184,22 +164,18 @@ describe('Media Query Transformer', () => {
...
     const expectedStyles = {
-      colorMode: {
-        mode: {
-          default: 'normal',
-          '@media (color) and (not (monochrome))': 'colorful',
-          '@media (monochrome)': 'grayscale',
-        },
+      mode: {
+        default: 'normal',
+        '@media (color) and (not (monochrome))': 'colorful',
```
Comment (nmn) [inline packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js:250]: Not a blocker for this PR, but note that this should be simplified in the future. color and monochrome should be considered mutually exclusive.

**Review by nmn — COMMENTED** [review]
Please add a few more examples to at least show that `min-width`, `max-width` and the combination of both is implemented. We can ignore the other edge-cases for now.

File: packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js (hunk 9 above)
Comment (mellyeliu) [reply packages/@stylexjs/babel-plugin/__tests__/transform-stylex-create-test.js:1835]: maybe `enableMediaQueryOrder`

File: packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js [hunk 13]
```diff
@@ -438,4 +421,30 @@ describe('Media Query Transformer', () => {
...
+          '@media (max-width: 1040px) and (min-width: 600px)': '1 / 3',
+        },
+      },
+    };
+
+    const expectedStyles = {
+      foo: {
+        gridColumn: {
+          default: '1 / 2',
+          '@media (min-width: 900px) and (max-width: 1440px) and (not ((min-width: 600px) and (max-width: 1040px)))':
```
Comment (mellyeliu) [inline packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js:436]: this still needs to be simplified, will handle separately

Comment (mellyeliu) [reply packages/style-value-parser/src/at-queries/__tests__/media-query-transform-test.js:436]: https://github.com/facebook/stylex/pull/1170

**Review by vincentriemer — APPROVED** [review]
